sdks/python/apache_beam/utils/thread_pool_executor.py

```diff
@@ -137,35 +101,33 @@ def submit(self, fn, *args, **kwargs):
     """
     future = _base.Future()
     work_item = _WorkItem(future, fn, args, kwargs)
-    try:
-      # Keep trying to get an idle worker from the queue until we find one
-      # that accepts the work.
-      while not self._idle_worker_queue.get(
-          block=False).accepted_work(work_item):
-        pass
-      return future
-    except queue.Empty:
-      with self._lock:
-        if self._shutdown:
-          raise RuntimeError(
-              'Cannot schedule new tasks after thread pool '
-              'has been shutdown.')
-
-        worker = _Worker(
-            self._idle_worker_queue,
-            self._permitted_thread_age_in_seconds,
-            work_item)
+    with self._lock:
+      if self._shutdown:
+        raise RuntimeError(
+            'Cannot schedule new tasks after thread pool has been shutdown.')
+      try:
+        self._idle_worker_queue.get(block=False).assign_work(work_item)
+
+        # If we have more idle threads then the max allowed, shutdown a thread.
+        if self._idle_worker_queue.qsize() > self._max_idle_threads:
+          try:
+            self._idle_worker_queue.get(block=False).shutdown()
```

Comment:
should we remove a total of `self._max_idle_threads - self._idle_worker_queue.qsize()` workers rather than just one?
IIUC, this is the only point (besides shutdown) where workers are removed, so maybe yes?

Reply:
I was thinking removing one at a time to reduce the rate at which we kill threads would "average" out the thread creation/death rate better but I have no data to support this hunch.

Reply:
Removing one thread at a time seems sensible to me.

Reply:
sounds good then!